protected/controllers/DayThreeController.php

```diff
@@ -0,0 +1,124 @@
+
+class DaythreeController extends Controller
+{
+
+    public function actionAbout()
+    {
+        $this->render('about');
+    }
+
+    public function actionError()
+    {
+        if ($error = Yii::app()->errorHandler->error) {
+            if (Yii::app()->request->isAjaxRequest) {
+                echo $error['message'];
+            } else {
+                $this->render('error', $error);
+            }
+        }
+    }
+
+    public function actionKaprekar()
+    {
+        $kaprekar_numbers = array();
+        $index            = 0;
+
+        for ($base_num = 1; $base_num < 1000; $base_num++) {
+            $square_num = pow($base_num, 2);
+            $square_str = (string)$square_num;
+            $mid        = (int)strlen($square_str) / 2;
+            $left       = (int)substr($square_str, 0, $mid);
+            $right      = (int)substr($square_str, $mid);
+            $sum        = $left + $right;
+
+            if ($sum == $base_num) {
+                $kaprekar_numbers[$index] = $sum;
+                $index++;
+            }
+        }
+        $this->render('kaprekar', array('kaprekar_numbers' => $kaprekar_numbers));
+    }
+
+    public function actionMatrix()
+    {
+        $matrix_model = new MatrixForm;
+
+        if (isset($_POST['MatrixForm'])) {
+            $matrix_model->attributes = $_POST['MatrixForm'];
+            $row_length               = $matrix_model->row_length;
+            $column_length            = $matrix_model->column_length;
+            $rotation                 = $matrix_model->rotation;
+            $matrix                   = array();
+            $count                    = 0;
+
+            for ($row = 0; $row < $row_length; ++$row) {
+                for ($column = 0; $column < $column_length; ++$column) {
+                    $count                 += 1;
+                    $matrix[$row][$column] = "<td>".$count."</td>";
+                }
+            }
+            $rotated_matrix = $this->actionCounterClockwise($column_length, $matrix, $row_length, $rotation);
+            $new_matrix     = $this->actionPrintMatrix($rotated_matrix);
+
+            $this->render('matrixtable', array('matrix' => $matrix, 'new_matrix' => $new_matrix));
+        } else {
+            $this->render('matrix', array('matrix_model' => $matrix_model));
+        }
+    }
+
+    public function actionCounterClockwise($column_length, $matrix, $row_length, $times_rotation)
+    {
+        $start_row    = 0;
+        $start_column = 0;
+
+        while ($row_length > 1 && $column_length > 1) {
+            $matrix_length   = ((2 * $row_length) + (2 * $column_length)) - 4;
+            $actual_rotation = $times_rotation % $matrix_length;
+
+            for ($rotation = 0; $rotation < $actual_rotation; $rotation++) {
+                $left_top_corner = $matrix[$start_row][$start_column];
+
+                for ($column_index = $start_column; $column_index < $start_column + $column_length - 1; $column_index++) {

```

Comment:
We have three nested loops, and there are 4 innermost loops. Try to avoid this, since this will be a usual candidate for refactoring or cleaning.